### ISSUE TITLE
Investigate slow parachains

### ISSUE THREAD

**Issue by ggwpez** [paritytech/polkadot-sdk#4569]
Discussions in the Fellowship chat did not lead to any actionable outcome. Not sure whom to ping here or what the expected behaviour is.

Polkadot parachains are not producing as many blocks as they should. You can use [this](https://github.com/ggwpez/substrate-scripts/blob/d4474f4766f4e71cca2f661ce863f93f077dda1e/plot-clock-skew.py) script to plot the skew (but it needs adjustments per chain).

The relay itself is fine:
![Screenshot%202024-05-22%20at%2015 04 45](https://github.com/paritytech/polkadot-sdk/assets/10380170/9155883e-2ef9-450f-bc38-6e44fce0a9c8)

But PAH is missing 30 days over the year:
![Screenshot%202024-05-22%20at%2014 47 49](https://github.com/paritytech/polkadot-sdk/assets/10380170/9195281f-51ed-48ba-9b8c-a062700f7414)

And collectives:
![Screenshot%202024-05-22%20at%2014 27 47](https://github.com/paritytech/polkadot-sdk/assets/10380170/db9a75e0-5781-4c4b-a63e-e8ab61531fe3)

**Comment by sandreim** [comment]
Yes indeed: https://grafana.teleport.parity.io/goto/TA1mEjPIR?orgId=1  some older data. Currently the introspector deployment is down, will be fixed tomorrow ASAP.

<img width="1418" alt="Screenshot 2024-05-26 at 20 06 31" src="https://github.com/paritytech/polkadot-sdk/assets/54316454/31756145-4a65-415b-a78f-4c56acbd14d3">

**Comment by sandreim** [comment]
Regarding why this is happening, AFAIR parachains not always get stuff backed, but I am expecting this to be way better when they switch to async backing.

**Comment by sandreim** [comment]
Here is one that did, but yeah, not rock solid. I think we should invest some time into understanding what is causing them to miss slots:

<img width="716" alt="Screenshot 2024-05-26 at 20 13 39" src="https://github.com/paritytech/polkadot-sdk/assets/54316454/5f5bf39d-a7a8-4c2f-811d-96665bbb15ae">

**Comment by alexggh** [comment]
>But PAH is missing 30 days over the year:

Using such a long period is tricky, because you have to exclude all past *solved* incidents on both polkadot and the collator set which would contribute to this skewing.

> I think we are still not at rock stable 12s blocks for parachains, which is what this plot also reveals.

Running on a network fully under our control where validators and collator have perfect connectivity get us to almost stable 12s and 6s(almost because availability cores are always cleared at session boundaries, so your going to miss a slot there). 

Our real networks are not in the perfect state, so blocks could be missed for a variety of reason(not saying this is what happens):
1.  Collators did not advertise their block in time.
2. Collators can't connect to validators in their backing group.(For example if I run https://github.com/lexnv/subp2p-explorer I only am able to reach around 86% of the validator set, Collators connecting to validators might have the same success rate)
```
  Discovered 287/331 authorities
```
3. Valid statements weren't gossiped in time and did not reach the block author.
4. One or more of the collators not producing blocks because of whatever reasons(it is behind or has some other issues)
5. Some unknown bugs or known bugs(https://github.com/paritytech/polkadot-sdk/issues/3590)

> Regarding why this is happening, AFAIR parachains not always get stuff backed (collations not getting to backing group fast enough or at all) , but I am expecting this to be way better when they switch to async backing.

Yes, switching to async backing should alleviate some of the problems above, but I don't think all of them.

> I think we should invest some time into understanding what is causing them to miss slots

I will try to allocate some time in the following weeks, but I wouldn't expect a quick resolution on this one.

**Comment by KarimJedda** [comment]
Some more context/data for this, hoping it's helpful:

## Polkadot BridgeHub

![Screenshot from 2024-06-27 12-28-27](https://github.com/paritytech/polkadot-sdk/assets/5029053/472f1123-1cd2-4591-8da5-b60e562ab551)


Took 24 seconds: https://bridgehub-polkadot.subscan.io/block/2883345 
Took 18 seconds: https://bridgehub-polkadot.subscan.io/block/2883344

## Kusama

![Screenshot from 2024-06-27 10-24-04](https://github.com/paritytech/polkadot-sdk/assets/5029053/0914608c-b6b8-49b8-8245-85e113b10be9)

Took 12 seconds: https://kusama.subscan.io/block/23792945

**Comment by burdges** [comment]
> Collators can't connect to validators in their backing group.(For example if I run https://github.com/lexnv/subp2p-explorer I only am able to reach around 86% of the validator set, Collators connecting to validators might have the same success rate)

Any idea why this happened?

It's likely prioritization on the validator side?  Any idea if all the validators are talking to one another though?  287/331 is way above what we'd notice in availability, but maybe we should've some tool that says "validator x never puts validator y into its availability bitfields"  Over a long enojugh time would say x cannot really talk to y, or at least their connection is very slow..

**Comment by alexggh** [comment]
> >Collators can't connect to validators in their backing group.(For example if I run https://github.com/lexnv/subp2p-explorer I only am able to reach around 86% of the validator set, Collators connecting to validators might have the same success rate)

> Any idea why this happened?

> It's likely prioritization on the validator side? Any idea if all the validators are talking to one another though? 287/331 is way above what we'd notice in availability.

When` I wrote that comment it was just a theory, but while looking at parity's collators for blocks in https://github.com/paritytech/polkadot-sdk/issues/4569#issuecomment-2194842358, there doesn't seem to be any connectivity issue, candidates are always advertised. Given the backing group is 5 on polkadot, I think we've got enough redundancy to alleviate any problems there.

Connecting to 287 out of 331 authorithies is what subp2p-explorer running on my local machine reports, the current theory I discussed with @lexnv  is that those are behind a NAT/Firewall and can't be reached from outside, the good thing is that does not mean they are not connected to other validators at all, because normally at each session change any two validators A & B will each try to connect to the other, so even if A can't reach B, B will be able to reach A.

Prioritization  could be an explanation as well.


> > "validator x never puts validator y into its availability bitfields" Over a long enojugh time would say x cannot really talk to y, or at least their connection is very slow..

Yeah, that's one of those things we should have tooling for and monitor if we want to reach the maximum parachain block throughput/

**Comment by bkchr** [comment]
> notified them or kicked them out of the active set,

If you can not proof that they have done any misbehavior, they will not get kicked out of the set. I mean you can create social pressure by having a list of shame, but that's it.

**Comment by alexggh** [comment]
> > notified them or kicked them out of the active set,

> If you can not proof that they have done any misbehavior, they will not get kicked out of the set. I mean you can create social pressure by having a list of shame, but that's it.

Yes, that's what I was thinking, when I said kick them out.
At the end of the day our protocols have the assumption that they work in the presence of 1/3 not doing their job or doing it badly, but that does not mean it will work at full throttle. Ideally, the points system would be hooked to all parts of our protocols and validators not performing would be kicked out, by not getting rewards and then they won't be nominated anymore, but even with that I still think we need automated tooling to pin point problems about the health of the network, so that you can help honest validators fix their nodes.

**Comment by alexggh** [comment]
Closing this is an old issue any problems here should be covered by the "Block confidence" projects, project view is here:
https://github.com/orgs/paritytech/projects/119/views/27
